lib/Dialect/TTNN/Analysis/LegalLayoutAnalysis.cpp

```diff
@@ -0,0 +1,293 @@
+        layout.withBufferType(op->getContext(), BufferType::L1)
+            .withElementType(op->getContext(), elementType);
+
+    assert(analysisInput.maxGrid.getShape().size() == 2 &&
+           "Max device grid is expected to be 2D.");
+    // Block Sharded
+    for (auto width = 1; width <= analysisInput.maxGrid.getShape()[0];
+         ++width) {
+      for (auto height = 1; height <= analysisInput.maxGrid.getShape()[1];
+           ++height) {
+        shardedResults.push_back(
+            shardedBase
+                .withGrid(op->getContext(), tensorType,
+                          GridAttr::get(op->getContext(), {width, height}))
+                .withMemoryLayout(op->getContext(),
+                                  TensorMemoryLayout::BlockSharded));
+      }
+    }
+
+    int64_t numCores = analysisInput.maxGrid.getGridVolume();
+    // Height Sharded
+    // TODO(odjuricic): Missing affine mapping to actual grid. Need to check
+    // with runtime implementation on what to produce here.
+    for (auto height = 1; height <= numCores; ++height) {
+      shardedResults.push_back(
+          shardedBase
+              .withGrid(op->getContext(), tensorType,
+                        GridAttr::get(op->getContext(), {height, 1}))
+              .withMemoryLayout(op->getContext(),
+                                TensorMemoryLayout::HeightSharded));
+    }
+
+    // Width Sharded
+    for (auto width = 1; width <= numCores; ++width) {
+      shardedResults.push_back(
+          shardedBase
+              .withGrid(op->getContext(), tensorType,
+                        GridAttr::get(op->getContext(), {1, width}))
+              .withMemoryLayout(op->getContext(),
+                                TensorMemoryLayout::WidthSharded));
+    }
+  }
+
+  // Filter layouts based on output tensor legality for current op.
+  shardedResults.erase(
+      std::remove_if(shardedResults.begin(), shardedResults.end(),
+                     [this](TTNNLayoutAttr layout) {
+                       return !tensor_shape_compatible_with_shard(op, layout) ||
+                              !mock_is_output_tensor_legal_for_op(op, layout);
+                     }),
+      shardedResults.end());
+
+  // Pick top largest sharded grids.
+  // This becomes a problem when we introduce row_major since an 8x8 tensor can
+  // be sharded onto a 8x8 grid.
+  std::sort(shardedResults.begin(), shardedResults.end(),
+            [](TTNNLayoutAttr a, TTNNLayoutAttr b) {
+              return a.getGrid().getGridVolume() > b.getGrid().getGridVolume();
+            });
+
+  analysisResult.insert(
+      analysisResult.end(), shardedResults.begin(),
+      shardedResults.begin() +
+          std::min(analysisInput.maxShardedGrids,
+                   static_cast<int64_t>(shardedResults.size())));
+
+  // Apply partial layout overrides.
+  if (override.has_value()) {
+    analysisResult.erase(
+        std::remove_if(analysisResult.begin(), analysisResult.end(),
```

Comment:
I do not like this approach. Can we at the beginning load all overrides and not even add layout not compatible with overrides instead of adding all and then removing them. At the very minimal create a util function instead of this lambda remove_if approach.

Reply:
I started with that approach and implemented a solution where i first create a search space for each layout param, then apply overrides and then try to itterate over the cros product and generate legal solutions. This was very hard to read because it ended up having a lot of nested if to check for non compatible cases across parameters. So i decided to go back to this which is cleaner an less error prone. I understand we are trading of performance for readability here, but it seems logical atm?

Will extract into a seperate funciton.